File: src/functions/detectPitchClassType.ts
/**
 * Automatically detects the format of pitch class data values from an array of string values.
 *
 * This function analyzes pitch class data values to determine its format, enabling automatic
 * parsing and conversion. It's essential for import workflows where the data format
 * might not be explicitly specified.
 *
 * The detection logic follows this priority:
 * 1. Fraction patterns (e.g., "3/2", "4:3")
 * 2. Fret division values (ascending numbers ending at 1000 or > 1200)
 * 3. Cents values (ascending numbers < 1200)
 * 4. Decimal ratios (ascending numbers 1.0-2.0)
 * 5. String lengths (descending numbers)
 *
 * @param values - Array of pitch class values as strings
 * @returns The detected format type or "unknown" if indeterminate
 *
 * @example
 * detectPitchClassType(["1/1", "9/8", "5/4", "4/3"]) // Returns "fraction"
 *
 * @example
 * detectPitchClassType(["0", "102", "201", "298", "1000"]) // Returns "fretDivision"
 *
 * @example
 * detectPitchClassType(["0", "204", "386", "498"]) // Returns "cents"
 *
 * @example
 * detectPitchClassType(["1.0", "1.125", "1.25", "1.333"]) // Returns "decimalRatio"
 *
 * @example
 * detectPitchClassType(["10000", "9600", "8700", "7000"]) // Returns "stringLength"
 */
export default function detectPitchClassValueType(values: string[]): "fraction" | "fretDivision" | "cents" | "decimalRatio" | "stringLength" | "unknown" {
  // Handle empty or invalid input
  if (!values.length) return "unknown";

  // Clean and filter the input values
  const cleaned = values.map((v) => v.trim()).filter((v) => v !== "");
  if (!cleaned.length) return "unknown";

  // Test for fraction patterns first (highest priority)
  // Matches patterns like "3/2", "4:3", "22/15"
  const fractionRegex = /^[1-9]\d*[/:][1-9]\d*$/;
  const allFractions = cleaned.every((val) => fractionRegex.test(val));

  if (allFractions) {
    // Verify fractions are in ascending order (expected for pitch scales)
    const fractionNumbers = cleaned.map((val) => {
      const [num, denom] = val.split(/[/:]/).map(Number);
      return num / denom;
    });

    // Check if values are in ascending order (required for valid scale)
    const isAscending = fractionNumbers.every((val, i, arr) => i === 0 || val >= arr[i - 1]);
    if (!isAscending) return "unknown";

    return "fraction";
  }

  // Parse all values as numbers for further analysis
  const numericVals = cleaned.map(Number);
  if (numericVals.some(isNaN)) return "unknown"; // Contains non-numeric values

  // Check if values are in ascending or descending order
  const ascending = numericVals.every((val, i, arr) => (i === 0 ? true : val >= arr[i - 1]));
  const descending = numericVals.every((val, i, arr) => (i === 0 ? true : val <= arr[i - 1]));

  // Test for fret division format (ascending values ending at 1000 or exceeding 1200)
  // Fret divisions are distinguishable from cents by either:
  // 1. Having a last value at or near 1000 (fixed end point)
  // 2. Having values that exceed 1200 (beyond one octave in cents)
  if (ascending) {
    const lastValue = numericVals[numericVals.length - 1];
    const hasFixedEndpoint = Math.abs(lastValue - 1000) < 50; // Within 50 of 1000
    const exceedsOctave = numericVals.some((v) => v > 1200);
    
    if (hasFixedEndpoint || exceedsOctave) {
      return "fretDivision";
    }
  }

  // Test for cents format (ascending values within 0-1200 range)
  if (ascending && numericVals.every((v) => v >= 0 && v < 1200)) {
    return "cents";
  }

  // Test for decimal ratio format (ascending values in typical ratio range)
  if (ascending && numericVals.every((v) => v >= 1.0 && v < 2.0)) {
    return "decimalRatio";
  }

  // Test for string length format (descending values)
  // String lengths typically decrease as pitch goes higher
  if (descending) {
    return "stringLength";
  }

  // Unable to determine format with confidence
  return "unknown";
}

File: python/functions/detect_pitch_class_type.py
"""
Detection utilities for determining the type/format of pitch class data.

This module provides functions to automatically detect whether pitch class data
is represented as fractions, cents, decimal ratios, or MIDI numbers.
"""

import re
from typing import Union


def detect_pitch_class_type(value: Union[str, int, float]) -> str:
    """
    Automatically detects the format type of a pitch class value.
    
    This function analyzes the input and determines whether it represents:
    - "fraction": Ratio notation like "3/2" or "5/4"
    - "cents": Cent values, typically between 0-1200
    - "decimal": Decimal ratio like 1.5 or 1.25
    - "midi": MIDI note numbers (integers 0-127)
    
    Args:
        value: The pitch class value to analyze
        
    Returns:
        String indicating the detected type: "fraction", "cents", "decimal", or "midi"
        
    Examples:
        >>> detect_pitch_class_type("3/2")
        "fraction"
        >>> detect_pitch_class_type(702.0)
        "cents"
        >>> detect_pitch_class_type(1.5)
        "decimal"
        >>> detect_pitch_class_type(60)
        "midi"
    """
    # Handle string inputs
    if isinstance(value, str):
        # Check if it's a fraction (contains '/')
        if "/" in value and re.match(r"^\d+/\d+$", value.strip()):
            return "fraction"
        
        # Try to convert to float for further analysis
        try:
            float_value = float(value)
        except ValueError:
            # If can't convert to number, assume it's some other string format
            return "fraction"  # Default assumption for strings
            
        # Continue analysis with the converted float
        value = float_value
    
    # Handle numeric inputs (int or float)
    if isinstance(value, (int, float)):
        # MIDI note numbers are typically integers between 0 and 127
        if isinstance(value, int) and 0 <= value <= 127:
            return "midi"
        
        # Cents are typically in the range 0-1200 (one octave)
        # But can be negative or larger for multiple octaves
        if isinstance(value, (int, float)) and -2400 <= value <= 2400:
            # Additional check: if it's a "round" decimal like 1.5, 2.0, etc.
            # it's more likely to be a decimal ratio than cents
            if isinstance(value, float) and 0.5 <= value <= 8.0:
                # Check if it's close to common musical ratios
                common_ratios = [1.0, 1.125, 1.2, 1.25, 1.333, 1.5, 1.6, 1.667, 1.8, 2.0]
                tolerance = 0.01
                
                for ratio in common_ratios:
                    if abs(value - ratio) < tolerance:
                        return "decimal"
            
            # If the value looks like typical cent values (often has decimal places)
            # or is in a range that's clearly cents
            if value > 8.0 or value < 0.5:
                return "cents"
        
        # For values clearly outside cent range but positive
        if value > 2400:
            return "cents" if value < 10000 else "decimal"
        
        # For small positive ratios
        if 0.5 <= value <= 8.0:
            return "decimal"
        
        # Default case for other numeric values
        return "cents"
    
    # Fallback for any other type
    return "fraction"


def is_fraction_format(value: str) -> bool:
    """
    Checks if a string is in fraction format (e.g., "3/2", "5/4").
    
    Args:
        value: String to check
        
    Returns:
        True if the string matches fraction format, False otherwise
        
    Examples:
        >>> is_fraction_format("3/2")
        True
        >>> is_fraction_format("1.5")
        False
        >>> is_fraction_format("702")
        False
    """
    if not isinstance(value, str):
        return False
    
    return bool(re.match(r"^\d+/\d+$", value.strip()))


def is_valid_cents_value(value: Union[str, int, float]) -> bool:
    """
    Checks if a value represents a valid cents measurement.
    
    Cents typically range from 0 to 1200 for one octave, but can extend
    beyond this range for multiple octaves or negative intervals.
    
    Args:
        value: Value to check
        
    Returns:
        True if the value appears to be a valid cents measurement
        
    Examples:
        >>> is_valid_cents_value(702.0)
        True
        >>> is_valid_cents_value(1200)
        True
        >>> is_valid_cents_value(0.5)
        False
    """
    try:
        numeric_value = float(value)
        # Allow a wide range for cents, including negative values and multiple octaves
        return -4800 <= numeric_value <= 4800
    except (ValueError, TypeError):
        return False


def is_valid_midi_note(value: Union[str, int, float]) -> bool:
    """
    Checks if a value represents a valid MIDI note number.
    
    MIDI notes range from 0 to 127, where 60 is middle C.
    
    Args:
        value: Value to check
        
    Returns:
        True if the value is a valid MIDI note number
        
    Examples:
        >>> is_valid_midi_note(60)
        True
        >>> is_valid_midi_note(128)
        False
        >>> is_valid_midi_note("60")
        True
    """
    try:
        numeric_value = int(float(value))
        return 0 <= numeric_value <= 127 and numeric_value == float(value)
    except (ValueError, TypeError):
        return False


def is_valid_decimal_ratio(value: Union[str, int, float]) -> bool:
    """
    Checks if a value represents a valid decimal ratio for musical intervals.
    
    Musical ratios are typically positive numbers, often between 0.5 and 8.0
    for common intervals within a few octaves.
    
    Args:
        value: Value to check
        
    Returns:
        True if the value appears to be a valid decimal ratio
        
    Examples:
        >>> is_valid_decimal_ratio(1.5)
        True
        >>> is_valid_decimal_ratio(2.0)
        True
        >>> is_valid_decimal_ratio(-1.0)
        False
    """
    try:
        numeric_value = float(value)
        return 0.25 <= numeric_value <= 16.0 and numeric_value > 0
    except (ValueError, TypeError):
        return False
